File: src/adapters/claude_code/schemas.ts
import { z } from 'zod';

const UsageSchema = z.object({
  input_tokens: z.number().int().nonnegative(),
  output_tokens: z.number().int().nonnegative(),
  cache_read_input_tokens: z.number().int().nonnegative().optional().default(0),
  cache_creation_input_tokens: z.number().int().nonnegative().optional().default(0),
  cache_creation: z.object({
    ephemeral_5m_input_tokens: z.number().int().nonnegative().optional().default(0),
    ephemeral_1h_input_tokens: z.number().int().nonnegative().optional().default(0),
  }).optional(),
  service_tier: z.string().optional(),
}).passthrough();

const ContentBlockSchema = z.discriminatedUnion('type', [
  z.object({ type: z.literal('text'), text: z.string() }).passthrough(),
  z.object({ type: z.literal('thinking'), thinking: z.string().optional() }).passthrough(),
  z.object({ type: z.literal('tool_use'), id: z.string(), name: z.string() }).passthrough(),
  z.object({ type: z.literal('tool_result') }).passthrough(),
]);

const BaseLineSchema = z.object({
  sessionId: z.string(),
  uuid: z.string(),
  timestamp: z.string(),
  cwd: z.string(),
  version: z.string().optional(),
  userType: z.string().optional(),
  entrypoint: z.string().optional(),
  gitBranch: z.string().optional(),
  isSidechain: z.boolean().optional(),
  agentId: z.string().optional(),
}).passthrough();

export const AssistantLineSchema = BaseLineSchema.extend({
  type: z.literal('assistant'),
  message: z.object({
    id: z.string(),
    model: z.string(),
    role: z.literal('assistant'),
    content: z.array(ContentBlockSchema).default([]),
    stop_reason: z.string().nullable().optional(),
    usage: UsageSchema,
  }).passthrough(),
  requestId: z.string().optional(),
  attribution_agent: z.string().optional(),
});

export const UserLineSchema = BaseLineSchema.extend({
  type: z.literal('user'),
  message: z.object({
    role: z.literal('user'),
    content: z.union([z.string(), z.array(z.any())]),
  }).passthrough(),
});

export type AssistantLine = z.infer<typeof AssistantLineSchema>;
export type UserLine = z.infer<typeof UserLineSchema>;

File: python/argus/adapters/claude_code/schemas.py
"""pydantic schemas for one line of a Claude Code JSONL session log.

Equivalents of the zod schemas in src/adapters/claude_code/schemas.ts.
``model_config = ConfigDict(extra='allow')`` mirrors zod's ``.passthrough()``
so unknown fields like ``attribution_agent`` survive parsing.
"""
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class _Passthrough(BaseModel):
    model_config = ConfigDict(extra="allow")


class CacheCreation(_Passthrough):
    ephemeral_5m_input_tokens: int = 0
    ephemeral_1h_input_tokens: int = 0


class Usage(_Passthrough):
    input_tokens: int = Field(ge=0)
    output_tokens: int = Field(ge=0)
    cache_read_input_tokens: int = Field(default=0, ge=0)
    cache_creation_input_tokens: int = Field(default=0, ge=0)
    cache_creation: CacheCreation | None = None
    service_tier: str | None = None


class AssistantMessage(_Passthrough):
    id: str
    model: str
    role: Literal["assistant"]
    content: list[dict[str, Any]] = Field(default_factory=list)
    stop_reason: str | None = None
    usage: Usage


class UserMessage(_Passthrough):
    role: Literal["user"]
    # Either a plain string ("hello") or a list of content blocks
    # ([{type: 'tool_result', ...}, ...]).
    content: str | list[Any]


class AssistantLine(_Passthrough):
    type: Literal["assistant"]
    sessionId: str
    uuid: str
    timestamp: str
    cwd: str
    version: str | None = None
    userType: str | None = None
    entrypoint: str | None = None
    gitBranch: str | None = None
    isSidechain: bool | None = None
    agentId: str | None = None
    requestId: str | None = None
    attribution_agent: str | None = None
    message: AssistantMessage


class UserLine(_Passthrough):
    type: Literal["user"]
    sessionId: str
    uuid: str
    timestamp: str
    cwd: str
    version: str | None = None
    userType: str | None = None
    entrypoint: str | None = None
    gitBranch: str | None = None
    isSidechain: bool | None = None
    agentId: str | None = None
    message: UserMessage
